Solution for list_posts_by_rating challenge

pmop opened the pull request:
This is my solution for the lists_posts_by_rating challenge.

I'm still not sure if I understood the ranking criteria correctly. The one I implemented ranks by highest upvote ratio first, then highest upvote-downvote.
Quoting from the [doc](https://github.com/weareswat/challenges/blob/master/1-development/list_posts_by_rating.md):
> Your solution should consider that if you have a first post with 600 up votes and 400 down votes means that you have 60% of up votes and 40% of down votes, and if you have another post with 6 up votes and 4 down votes you have the same % of the previous post, 60% of up votes and 40% of down votes.
> Hint: Note that the score in % of the two posts are equal, but the real "values" are significantly differents

What I took for it is that if the upvote ratio is the same for both, then the one with more upvotes is more "relevant" and hence should be ranked higher.

However the criteria I came up with will put 100% upvote ratio on top, so there can be a scenario where there's one post with few upvotes but all upvotes, so 100% upvote ratio, ranking higher than posts with less upvote ratio but more upvotes overall. Ex: a Post with 1 upvote, 0 downvotes, will rank higher than one with 60/40.

**Let me know if I misunderstood the criteria and if I can change it.**

Other than that, few other comments:

1st solution (prior to commit 06bbfb9) just grabbed all posts and sorted. No pagination.

The 2nd solution and the one I chose to deliver includes an index and uses ORDER BY clause in hopes of being more performant, it also uses the Pagy gem for pagination. I haven't bench-marked yet but my theory is that if you can use a B-TREE algo, you could traverse the tree and get the Posts, already ordered rank in descending, without having to do a sort, which happens to be the case for the index. It'd also help with keeping the posts ranked correctly when using pagination. I haven't benchmarked/tested it yet.


### Final comments
I'd use `/posts/:id/upvote` `/posts/:id/downvte` routes instead that you can do with `members` in `resources` since they're actions on posts.

Commit:
Project bootstrap

Commit:
Fix

Commit:
Scaffold Post

Commit:
Tests for manipulating upvotes/downvotes

Commit:
Implement methods for manipulating upvotes/downvotes

Commit:
Add tests for ratio method

Commit:
Add ratio method implementation

Commit:
Add test for score method

Commit:
Test sort criteria

Commit:
Implement sort with custom criteria

Commit:
Set upvote/downvote defaults

Commit:
Schema

Commit:
Rename scope

Commit:
Minor test refactor

Commit:
Add tests for some fixes

Commit:
Add fixtures

Commit:
Fix issues

Commit:
Add tests for /upvote

Commit:
Upvote action implementation

Commit:
Upvote and downvote routes

Commit:
Tests for downvote route

Commit:
Implement downvote action

Commit:
Merge branch 'feature/controller-and-routes' into feature/list_posts_by_rating_solution
Add tests for routes and implement routes and controller actions for upvote and downvote.

Commit:
Add README

Commit:
Clean up

Commit:
Rename some methods

Commit:
Clean up, add code coverage, improve tests

Commit:
Add db seed and improve the readme

Commit:
Avoid race conditions

Commit:
Minor refactor

Commit:
Add index for fast DB sort with ranking as criteria

Commit:
Add pagination and use db order clause by default

Use the new DB order clause that together with the index may bring perf improvements.
Use the Pagy gem for pagination.

Commit:
Add Pagy config

Commit:
Merge branch 'patch/use-index-for-sort' into feature/list_posts_by_rating_solution

Add index that takes in both rating and expression for upvotes - downvotes.
Use ORDER BY clause to leverage the index for a faster sort.
Add Pagy to paginate large results.

Commit:
Refactor

Make the code more DRY. Fix some issues.

Commit:
Setup logging

Commit:
Apply rubocop fixes

pmop commented:
Pretty much done at this point. I plan to do a little benchmarking and some testing before delivering.

Commit:
Debloat

Commit:
Small changes

Commit:
Improve the README

goncalopc96 commented on list_posts_by_rating/db/seeds.rb:
```diff
@@ -0,0 +1,16 @@
+#   movies = Movie.create([{ name: "Star Wars" }, { name: "Lord of the Rings" }])
+#   Character.create(name: "Luke", movie: movies.first)
+
+Post.create(upvotes: 600, downvotes: 0, rating: 100, content: 'I should rank 1st')
+Post.create(upvotes: 600, downvotes: 400, rating: 60, content: 'I should rank 2nd')
+Post.create(upvotes: 60, downvotes: 40, rating: 60, content: 'I should rank 3rd')
+Post.create(upvotes: 0, downvotes: 0, rating: 0, content: 'I should rank 4th')
+Post.create(upvotes: 1, downvotes: 2, rating: 0, content: 'I should rank 5th')
+Post.create(upvotes: 0, downvotes: 2, rating: 0, content: 'I should rank 6th')
```
O rating nao deveria ser automaticamente calculado atraves do numero de upvotes e downvotes?

pmop replied:
Ele é automaticamente calculado toda vez que é feito um incremento de upvote/downvote. Como na criação você tem 0 upvotes/downvotes, e o rating é uma divisão, ficaria 0/0, então na criação não faz sentido calcular. Mas eu entendo que seria possível fazer isso via hook e sempre calcular o rating quando o upvote/downvote for alterado.

goncalopc96 commented on list_posts_by_rating/app/controllers/posts_controller.rb:
```diff
@@ -0,0 +1,53 @@
+  private
+
+  def set_post
+    @post = Post.find(params[:id])
```
Se nao existir o respetivo post, o que acontece?

pmop replied:
Vai lançar uma exceção `ActiveRecord::RecordNotFound` e a resposta será um erro 404.

goncalopc96 commented on list_posts_by_rating/app/controllers/posts_controller.rb:
```diff
@@ -0,0 +1,53 @@
+    else
+      log_error(@post, vote_type)
+
+      render json: @post.errors, status: 500
```
Nao ha possibilidade de outros erros? Apenas 500?

pmop replied:
Talvez tenha e eu não estou pensando agora.
Na minha visão p/ erros internos, ou seja não foram causados pelo cliente, o 500 pode ser um guarda-chuva. A gente loga e tenta resolver o erro interno.
Da família de erros do cliente, aqui só faz sentido o not found, que vai ser tratado automaticamente quando o ActiveRecord lançar um `ActiveRecord::RecordNotFound`. Como o cliente não pode criar ou atualizar nada passando parâmetro, não faz sentido um bad request.

goncalopc96 commented on list_posts_by_rating/app/models/post.rb:
```diff
@@ -0,0 +1,32 @@
+  #
+  # @param vote_type [Symbol] The kind of vote to increment. Either :upvotes or :downvotes.
+  def increment_vote(vote_type)
+    new_votes_count = send(vote_type) + 1
```
porque nao usar o increment?

pmop replied:
Não conhecia esse método. Mudei p/ usar ele e também simplifiquei o código e mudei p/ usar o um hook também:  bb3c2d01c75c8a227ee34b32222097e1dee21eac

jeandias commented on list_posts_by_rating/app/models/post.rb:
```diff
@@ -0,0 +1,32 @@
+
+class Post < ApplicationRecord
+  scope :all_posts_ranked_desc, lambda {
+    all.order(Arel.sql('rating DESC, (upvotes - downvotes) DESC'))
```
porque usar Arel para escrever esse SQL?
qual a diferença entre `(upvotes - downvotes)` ou `upvotes`?

pmop replied:
Sem o `Arel`, ele lança um `ActiveRecord::UnknownAttributeReference` alertando `potentially unsafe value is passed to a query method`. 
O `upvotes - downvotes` é para poder ordenar os posts os quais downvotes > upvotes, para ordenar pelo menos pior ex: um Post com 1 upvote 2 downvotes (-1) é "menos pior" do que um post com  0 upvotes e 2 downvotes (-2).

jeandias commented on list_posts_by_rating/app/models/post.rb:
```diff
@@ -0,0 +1,32 @@
+
+    return 0 if score.negative?
+
+    (100 * (new_upvotes.to_f / votes)).to_i
```
transformar num inteiro não tira a possibilidade de um post com 50.5 estar numa posição melhor do que outro que tem 50.4?

pmop replied:
Nesse caso os dois ficariam 50 e o fator decisivo seria `upvotes - downvotes`.

jeandias commented on list_posts_by_rating/db/migrate/20230824202937_add_rating_index_to_posts.rb:
```diff
@@ -0,0 +1,7 @@
+
+class AddRatingIndexToPosts < ActiveRecord::Migration[7.0]
+  def change
+    add_index :posts, 'rating DESC, (upvotes - downvotes) DESC', name: 'index_posts_rating'
```
porque `(upvotes - downvotes)` e não simplesmente `upvotes`?

Commit:
Review changes

- Use `increment` method
- Simplify the code
- Use hooks for updating the rating
- Update the db seed

Commit:
Small change
